File: runner/internal/adapter/codex.go
package adapter

import (
	"fmt"
	"strings"
	"time"
)

// CodexUsageURL 可在测试中覆盖。
var CodexUsageURL = "https://chatgpt.com/backend-api/wham/usage"

// codex 适配器：Codex 订阅额度（chatgpt.com/backend-api/wham/usage，非官方）。
// 凭证：~/.codex/auth.json 的 tokens.access_token（ChatGPT 订阅 OAuth token，有效期约一周，过期需重新粘贴）。
// 可选 cred["base_url"]：正向转发地址，替换默认的 https://chatgpt.com/backend-api。
// 注意：实测必须带 codex CLI 风格的 User-Agent，否则被拦返回 403 HTML 错误页（不是 401）。
// 刻意不做 refresh_token 自动续期：refresh token 是一次性轮换的，刷新会使本机 Codex CLI 登录态作废。
type codexAdapter struct{}

func (codexAdapter) Provider() string { return "codex" }

type codexWindow struct {
	UsedPercent        *float64 `json:"used_percent"`
	LimitWindowSeconds float64  `json:"limit_window_seconds"`
	ResetAt            *float64 `json:"reset_at"` // unix 秒
}

type codexPayload struct {
	PlanType  string `json:"plan_type"`
	RateLimit *struct {
		Primary   *codexWindow `json:"primary_window"`
		Secondary *codexWindow `json:"secondary_window"`
	} `json:"rate_limit"`
	Credits *struct {
		HasCredits bool `json:"has_credits"`
		Balance    any  `json:"balance"` // 字符串或数字
	} `json:"credits"`
}

func (codexAdapter) Fetch(cred map[string]string) ([]Row, error) {
	token := cred["access_token"]
	if token == "" {
		token = cred["token"]
	}
	if token == "" {
		return nil, fmt.Errorf("missing credential: access_token（~/.codex/auth.json 的 tokens.access_token）")
	}
	headers := map[string]string{
		"Authorization": "Bearer " + token,
		"Accept":        "application/json",
		"originator":    "codex_cli_rs",
		"User-Agent":    "codex_cli_rs/0.40.0",
	}
	if id := cred["account_id"]; id != "" {
		headers["ChatGPT-Account-Id"] = id
	}
	var payload codexPayload
	url := CodexUsageURL
	if base := cred["base_url"]; base != "" {
		url = strings.TrimRight(base, "/") + "/wham/usage"
	}
	if err := getJSON(url, headers, &payload); err != nil {
		return nil, fmt.Errorf("codex usage（access_token 可能已过期，需重新粘贴）: %w", err)
	}

	var primary, secondary *codexWindow
	if payload.RateLimit != nil {
		primary, secondary = payload.RateLimit.Primary, payload.RateLimit.Secondary
	}
	if primary == nil && secondary == nil {
		return nil, fmt.Errorf("codex: no rate_limit window in response")
	}

	// 窗口身份靠 limit_window_seconds 判断（部分套餐周限额在 primary_window），
	// 缺时长时按 primary=5 小时 / secondary=周 兜底
	var rows []Row
	for _, w := range []struct {
		win      *codexWindow
		fallback string
	}{
		{primary, "session_used_pct"},
		{secondary, "weekly_used_pct"},
	} {
		if w.win == nil || w.win.UsedPercent == nil {
			continue
		}
		metric := w.fallback
		if secs := w.win.LimitWindowSeconds; secs >= 86400 {
			metric = "weekly_used_pct"
		} else if secs > 0 {
			metric = "session_used_pct"
		}
		rows = append(rows, Row{
			Provider:   "codex",
			Metric:     metric,
			Value:      *w.win.UsedPercent,
			Unit:       strptr("percent"),
			LimitValue: fptr(100),
			ResetAt:    unixISO(w.win.ResetAt),
		})
	}

	// 充值余额（可选字段，balance 可能是字符串）
	if c := payload.Credits; c != nil && c.HasCredits && c.Balance != nil {
		rows = append(rows, Row{
			Provider: "codex",
			Metric:   "credits_usd",
			Value:    num(c.Balance),
			Unit:     strptr("usd"),
		})
	}
	return rows, nil
}

// unixISO 把 unix 秒转 ISO 字符串；nil 返回 nil。
func unixISO(ts *float64) *string {
	if ts == nil {
		return nil
	}
	s := time.Unix(int64(*ts), 0).UTC().Format(time.RFC3339)
	return &s
}

File: cloudflare-hub/src/runner/adapters/codex.ts
import type { QuotaRow, QuotaAdapter } from "../types";

interface CodexWindow {
  used_percent?: number;
  limit_window_seconds?: number;
  reset_after_seconds?: number;
  reset_at?: number; // unix 秒
}

/**
 * chatgpt.com /backend-api/wham/usage（Codex 订阅额度，非官方；接口变动时记录 scrape_error）。
 * 凭证：~/.codex/auth.json 的 tokens.access_token（ChatGPT 订阅 OAuth token，有效期约一周，过期需重新粘贴）。
 * 可选 cred.base_url：正向转发地址（替换 https://chatgpt.com/backend-api，用于绕开对端 WAF 对 Workers 出口的拦截）。
 * 注意：必须带 codex CLI 风格的 User-Agent——实测默认 UA 被拦返回 403 HTML 错误页（不是 401）。
 * 刻意不做 refresh_token 自动续期：refresh token 是一次性轮换的，runner 刷新会使本机 Codex CLI 登录态作废。
 */
export const codex: QuotaAdapter = {
  provider: "codex",
  async fetch(cred, f) {
    const token = cred.access_token ?? cred.token;
    if (!token) throw new Error("missing credential: access_token（~/.codex/auth.json 的 tokens.access_token）");
    const headers: Record<string, string> = {
      Authorization: `Bearer ${token}`,
      Accept: "application/json",
      originator: "codex_cli_rs",
      "User-Agent": "codex_cli_rs/0.40.0",
    };
    if (cred.account_id) headers["ChatGPT-Account-Id"] = cred.account_id;

    const base = (cred.base_url ?? "https://chatgpt.com/backend-api").replace(/\/+$/, "");
    const res = await f(`${base}/wham/usage`, { headers });
    if (!res.ok) throw new Error(`codex usage: HTTP ${res.status}（access_token 可能已过期，需重新粘贴）: ${(await res.text()).slice(0, 200)}`);
    const json = (await res.json()) as {
      plan_type?: string;
      rate_limit?: { primary_window?: CodexWindow | null; secondary_window?: CodexWindow | null };
      credits?: { has_credits?: boolean; balance?: string | number };
    };

    const primary = json.rate_limit?.primary_window ?? null;
    const secondary = json.rate_limit?.secondary_window ?? null;
    if (!primary && !secondary) throw new Error("codex: no rate_limit window in response");

    // 窗口身份靠 limit_window_seconds 判断（部分套餐周限额在 primary_window），
    // 缺时长时按 primary=5 小时 / secondary=周 兜底
    const rows: QuotaRow[] = [];
    for (const [w, fallback] of [
      [primary, "session_used_pct"],
      [secondary, "weekly_used_pct"],
    ] as const) {
      if (!w || w.used_percent == null) continue;
      const secs = Number(w.limit_window_seconds ?? 0);
      const metric = secs >= 86400 ? "weekly_used_pct" : secs > 0 ? "session_used_pct" : fallback;
      rows.push({
        provider: "codex",
        metric,
        value: Number(w.used_percent),
        unit: "percent",
        limit_value: 100,
        reset_at: w.reset_at ? new Date(Number(w.reset_at) * 1000).toISOString() : null,
      });
    }

    // 充值余额（可选字段，balance 可能是字符串）
    const balance = Number(json.credits?.balance);
    if (json.credits?.has_credits && Number.isFinite(balance)) {
      rows.push({
        provider: "codex",
        metric: "credits_usd",
        value: balance,
        unit: "usd",
        limit_value: null,
        reset_at: null,
      });
    }
    return rows;
  },
};
